# How Does Global AI Adoption Rate Compare with Productivity Change? — Phase II
### ENGE707: Data Engineering and Machine Learning Assignment — LCS-Based Prediction

Suemon Kwok (14883335), Quentin Masoe (23191960), Avathanshu Bhat (23227963)

**Dataset:** Global AI Adoption and Workforce Impact Dataset (`ai_company_adoption.csv`)

This notebook continues on from the Phase I scoping report and works through the Phase II tasks (LCS baseline, preprocessing, improved LCS system, experiments, model comparison, interpretation, and discussion).

In [1]:
import os
import zipfile

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

pd.set_option("display.max_columns", 50)

## Task 1 - Project Phase I Summary

### 1.1 Problem restatement

- **Problem:** predict a company's reported productivity change from its AI adoption behaviour.
- **Target variable:** `productivity_change_percent` (continuous, %) — this is a **regression** problem.
- **Predictor (Phase I scope):** `ai_adoption_rate` (% of operations/processes using AI).
- **Prediction objective:** given a company-quarter's AI adoption rate (and, from Phase II onward, other available features), estimate the productivity change the company is likely to report.

In [2]:
# Load the dataset (extract from archive.zip first if the CSV isn't already alongside this notebook)
CSV_NAME = "ai_company_adoption.csv"
ZIP_NAME = "archive.zip"

if not os.path.exists(CSV_NAME) and os.path.exists(ZIP_NAME):
    with zipfile.ZipFile(ZIP_NAME) as z:
        z.extractall(".")

df = pd.read_csv(CSV_NAME)
df.shape

(150000, 43)

### 1.2 Dataset source, size, and characteristics

In [3]:
# Quick structural recap carried over from Phase I
print(f"Rows: {df.shape[0]:,}")
print(f"Columns: {df.shape[1]}")
print(f"Distinct companies: {df['company_id'].nunique():,}")
print(f"Countries: {df['country'].nunique()}")
print(f"Industries: {df['industry'].nunique()}")
print(f"Survey period: {df['survey_year'].min()}-{df['quarter'].min()} to {df['survey_year'].max()}-{df['quarter'].max()}")

Rows: 150,000
Columns: 43
Distinct companies: 10,000
Countries: 30
Industries: 9
Survey period: 2023-Q1 to 2026-Q4


In [4]:
# Dtype breakdown
df.dtypes.value_counts()

float64    15
int64      14
str        14
Name: count, dtype: int64

- **Source:** Global AI Adoption and Workforce Impact Dataset (Kaggle, CC0 licence).
- **Size:** 150,000 company-quarter survey responses from 10,000 distinct companies, across 30 countries and 9 industries, surveyed quarterly Q1 2023 – Q4 2026.
- **Features:** 43 columns total — a mix of numeric (AI investment, budget, maturity, satisfaction, financial metrics), categorical (industry, country, company size, tool used), and identifier columns.
- **Phase I focus variables:** `ai_adoption_rate` and `productivity_change_percent` only, per the narrowed scope agreed after Phase I lecturer feedback.

### 1.3 Main data-quality issues identified in Phase I

In [5]:
# Re-confirm the Phase I data-quality findings for the two focus variables
print("Missing values:")
print(df[["ai_adoption_rate", "productivity_change_percent"]].isna().sum())

print("\nDuplicate rows:", df.duplicated().sum())
print("Duplicate response_id:", df["response_id"].duplicated().sum())

print("\nai_adoption_rate range:", df["ai_adoption_rate"].min(), "-", df["ai_adoption_rate"].max())
print("productivity_change_percent range:", df["productivity_change_percent"].min(), "-", df["productivity_change_percent"].max())

Missing values:
ai_adoption_rate               0
productivity_change_percent    0
dtype: int64



Duplicate rows: 0
Duplicate response_id: 0

ai_adoption_rate range: 0.0 - 100.0
productivity_change_percent range: 0.0 - 34.36


- No missing values and no duplicate records were found in either focus variable.
- Both variables sit within their valid documented ranges (0–100% and 0–34.36% respectively), so no impossible values needed correcting.
- IQR-based outlier checks flagged well under 1% of records in each variable (504 and 529 respectively) — these were treated as genuine high-adoption/high-productivity companies, not data-entry errors.
- The dataset is an **unbalanced panel**: 10,000 companies each contribute an unequal number of quarterly rows, so the 150,000 records are not fully independent observations.
- A secondary rise near 0–2% in the productivity distribution suggested a distinct "zero/low-productivity" subgroup, later linked to lower mean AI adoption (18% vs. 36% overall).

### 1.4 Why this problem is suitable for an LCS-based solution

- The target (`productivity_change_percent`) is a **continuous outcome driven by conditions across several company attributes** (adoption behaviour, workforce factors, financial context) — LCS systems are well suited to problems where the relationship is expressed as a set of localised if-then rules over feature conditions, rather than a single global function.
- The dataset is **tabular with mixed numeric and categorical features**, which matches the input format LCS implementations (e.g. eLCS, XCS) are designed for.
- Phase I already showed the AI-adoption/productivity relationship is **not purely linear or global** — the zero-productivity subgroup behaved differently from the rest of the sample. LCS's rule-based, niche-covering approach can capture this kind of subgroup structure explicitly, and its rules remain human-readable, which supports the interpretability requirement in Task 7.
- The moderate size of the dataset (150,000 rows, reducible to a manageable feature set) is within the range LCS implementations such as eLCS can realistically train on.

### Note on Task 2 scope (resolved)

Confirmed direction: keep the Phase I two-variable scope (`ai_adoption_rate` -> `productivity_change_percent`) for the **original LCS baseline** in Task 2, since the other 41 columns were already dropped during the Phase I cleaning step and never carried forward. Any feature expansion, if used, will be introduced and justified as part of the **improved system** in Tasks 3/4, not the baseline.

## Task 2 - Original LCS System on Raw Dataset

### 2.1 LCS variant selection

**eLCS**, via the `scikit-eLCS` package (`skeLCS`) — a scikit-learn-compatible implementation of the educational Learning Classifier System (eLCS), exposing the standard `fit` / `predict` / `score` interface.

In [6]:
# scikit-eLCS / statsmodels aren't default Jupyter/Colab packages, install if missing
try:
    import skeLCS, statsmodels  # noqa: F401
except ImportError:
    %pip install -q scikit-eLCS statsmodels

from skeLCS import eLCS
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import (
    accuracy_score,
    balanced_accuracy_score,
    f1_score,
    classification_report,
)
from statsmodels.stats.contingency_tables import mcnemar

### 2.2 Minimal processing required for LCS compatibility

`skeLCS.eLCS` requires:
- all feature values to be numeric (or `NaN`),
- the target (`phenotype`) to be numeric **and discrete** — this implementation always treats the phenotype as a classification target, so a continuous variable cannot be passed directly.

`ai_adoption_rate` is already numeric, so no change is needed there. `productivity_change_percent`, however, is continuous, so it is minimally processed by binning it into three quantile-based classes (Low / Medium / High) and label-encoding them as integers 0/1/2. This is the smallest possible change that makes the existing Phase I target compatible with the LCS code — no algorithmic modification is made to eLCS itself.

In [7]:
# Minimal processing: discretise the continuous target into 3 balanced classes so it is
# compatible with eLCS's classification-only phenotype requirement
df["productivity_class"] = pd.qcut(
    df["productivity_change_percent"], q=3, labels=[0, 1, 2]
).astype(int)

df["productivity_class"].value_counts().sort_index()

productivity_class
0    50023
1    50031
2    49946
Name: count, dtype: int64

In [8]:
# Baseline feature set: the single Phase I predictor, kept unmodified
X = df[["ai_adoption_rate"]].values.astype(float)
y = df["productivity_class"].values

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

X_train.shape, X_test.shape

((120000, 1), (30000, 1))

### 2.3 LCS parameters

Default `skeLCS.eLCS` hyperparameters are used for the baseline (no tuning at this stage — tuning is reserved for the improved system in Task 4):

| Parameter | Value | Meaning |
|---|---|---|
| `learning_iterations` | 10,000 | number of training iterations (rule-learning cycles) |
| `N` | 1,000 | maximum micro-classifier population size |
| `nu` | 5 | fitness pressure exponent |
| `chi` | 0.8 | crossover probability |
| `mu` | 0.04 | mutation probability |
| `p_spec` | 0.5 | probability of specifying an attribute on covering |
| `discrete_attribute_limit` | 10 | max unique values before an attribute is treated as continuous |
| `random_state` | 42 | fixed for reproducibility |

In [9]:
# Train the original, unmodified eLCS baseline
elcs_baseline = eLCS(learning_iterations=10000, random_state=42)
elcs_baseline.fit(X_train, y_train)
print("eLCS baseline trained.")

eLCS baseline trained.


In [10]:
# Baseline performance (eLCS prediction is pure-Python and slow at scale, so train
# accuracy is estimated on a random subsample rather than all 120,000 training rows)
rng = np.random.default_rng(42)
sample_idx = rng.choice(len(X_train), size=5000, replace=False)

train_acc = elcs_baseline.score(X_train[sample_idx], y_train[sample_idx])
y_pred_test = elcs_baseline.predict(X_test)
test_acc = accuracy_score(y_test, y_pred_test)
coverage = elcs_baseline.get_final_instance_coverage()

print(f"Train accuracy (5,000-row sample): {train_acc:.4f}")
print(f"Test accuracy: {test_acc:.4f}")
print(f"Final instance coverage: {coverage:.4f}")

Train accuracy (5,000-row sample): 0.5387
Test accuracy: 0.5397
Final instance coverage: 1.0000


In [11]:
print(classification_report(y_test, y_pred_test, target_names=["Low", "Medium", "High"]))

              precision    recall  f1-score   support

         Low       0.61      0.72      0.66     10005
      Medium       0.00      0.00      0.00     10006
        High       0.50      0.90      0.64      9989

    accuracy                           0.54     30000
   macro avg       0.37      0.54      0.43     30000
weighted avg       0.37      0.54      0.43     30000



/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1833: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1833: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1833: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])


### 2.4 Baseline results summary

Random-guess accuracy for 3 balanced classes is ~33%. The single-feature eLCS baseline should be compared against this floor once results are in — a low margin above chance would support broadening the feature set (or improving preprocessing) as the Task 3/4 improvement, without changing the Task 2 baseline scope itself.

## Task 3 - Data Preprocessing and Feature Engineering

In [12]:
# Feature set for the improved system (domain-selected, avoids leakage columns like
# jobs_displaced / revenue_growth_percent which are downstream outcomes, not predictors)
selected_cols = [
    'ai_adoption_rate',
    'ai_maturity_score',
    'ai_failure_rate',
    'task_automation_rate',
    'ai_adoption_stage',
    'industry',
    'ai_training_hours',
]

df_model = df[selected_cols + ['productivity_change_percent']].copy()
df_model.isna().sum()

ai_adoption_rate               0
ai_maturity_score              0
ai_failure_rate                0
task_automation_rate           0
ai_adoption_stage              0
industry                       0
ai_training_hours              0
productivity_change_percent    0
dtype: int64

In [13]:
# No missing values, no duplicates, no out-of-range values in this subset (re-confirm)
print("Duplicates:", df_model.duplicated().sum())
df_model.describe(include='all').T

Duplicates: 0


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
ai_adoption_rate,150000.0,NaN,NaN,NaN,36.414726,14.544096,0.0,26.43,36.32,46.2,100.0
ai_maturity_score,150000.0,NaN,NaN,NaN,0.344675,0.13382,0.0,0.25,0.341,0.435,0.904
ai_failure_rate,150000.0,NaN,NaN,NaN,25.774337,7.761027,0.0,20.27,25.88,31.52,40.0
task_automation_rate,150000.0,NaN,NaN,NaN,19.761744,9.222562,0.0,13.36,19.7,26.02,59.63
ai_adoption_stage,150000,4,partial,78800,NaN,NaN,NaN,NaN,NaN,NaN,NaN
industry,150000,9,Technology,21107,NaN,NaN,NaN,NaN,NaN,NaN,NaN
ai_training_hours,150000.0,NaN,NaN,NaN,27.207475,13.136749,0.0,18.07,27.06,36.13,80.0
productivity_change_percent,150000.0,NaN,NaN,NaN,9.266996,5.637067,0.0,5.07,9.06,13.12,34.36


In [14]:
# Encode ai_adoption_stage as an ordinal variable (natural progression none < pilot < partial < full)
stage_order = {'none': 0, 'pilot': 1, 'partial': 2, 'full': 3}
df_model['ai_adoption_stage_enc'] = df_model['ai_adoption_stage'].map(stage_order)

# One-hot encode industry (nominal, no natural order)
df_model = pd.get_dummies(df_model, columns=['industry'], prefix='industry')
df_model = df_model.drop(columns=['ai_adoption_stage'])
df_model.head()

,ai_adoption_rate,ai_maturity_score,ai_failure_rate,task_automation_rate,ai_training_hours,productivity_change_percent,ai_adoption_stage_enc,industry_Agriculture,industry_Consulting,industry_Education,industry_Finance,industry_Healthcare,industry_Logistics,industry_Manufacturing,industry_Retail,industry_Technology
0,30.57,0.224,35.22,19.47,20.94,2.65,1,False,False,True,False,False,False,False,False,False
1,27.25,0.174,24.65,26.24,23.10,5.77,1,False,False,True,False,False,False,False,False,False
2,31.54,0.266,36.51,22.24,12.29,6.94,1,False,False,True,False,False,False,False,False,False
3,11.02,0.123,33.30,2.87,11.31,5.32,1,False,False,True,False,False,False,False,False,False
4,33.39,0.363,19.83,18.44,35.17,6.32,1,False,False,True,False,False,False,False,False,False


In [15]:
# Target: same 3-class quantile binning used for the Task 2 baseline, kept consistent
# for a fair comparison; class balance confirmed below (avoids introducing new imbalance)
df_model['productivity_class'] = pd.qcut(
    df_model['productivity_change_percent'], q=3, labels=[0, 1, 2]
).astype(int)

df_model['productivity_class'].value_counts(normalize=True).sort_index()

productivity_class
0    0.333487
1    0.333540
2    0.332973
Name: proportion, dtype: float64

In [16]:
# Final feature matrix / target for the preprocessed + feature-engineered dataset
feature_cols = [c for c in df_model.columns if c not in ['productivity_change_percent', 'productivity_class']]

X_fe = df_model[feature_cols].astype(float).values
y_fe = df_model['productivity_class'].values

X_fe.shape, y_fe.shape, feature_cols

((150000, 15),
 (150000,),
 ['ai_adoption_rate',
  'ai_maturity_score',
  'ai_failure_rate',
  'task_automation_rate',
  'ai_training_hours',
  'ai_adoption_stage_enc',
  'industry_Agriculture',
  'industry_Consulting',
  'industry_Education',
  'industry_Finance',
  'industry_Healthcare',
  'industry_Logistics',
  'industry_Manufacturing',
  'industry_Retail',
  'industry_Technology'])

No scaling was applied: eLCS matches continuous attributes by range internally, and the Decision Tree used later in the model comparison is scale-invariant, so scaling would add a processing step without changing either model's behaviour.

## Task 4 - Improved LCS-Based System

In [17]:
# Improvements vs. Task 2 baseline: (1) richer feature set (7 domain features instead of 1),
# (2) larger rule population to cover the bigger condition space, (3) correct-set subsumption
# switched on for rule generalisation/compaction
elcs_improved = eLCS(
    learning_iterations=10000,
    N=1500,
    do_correct_set_subsumption=True,
    random_state=42,
)

## Task 5 - Experiments

### 5.1 Validation strategy

80/20 train-test split, stratified on the target class. Chosen over k-fold cross-validation because the dataset is large (150,000 rows) and a single held-out split already gives a stable, low-variance estimate at a fraction of the compute cost of k-fold — the same split is reused across all models below for a fair, paired comparison.

In [18]:
X_train, X_test, y_train, y_test = train_test_split(
    X_fe, y_fe, test_size=0.2, random_state=42, stratify=y_fe
)
X_train.shape, X_test.shape

((120000, 15), (30000, 15))

### 5.2 Train all models on the same split

In [19]:
# Original LCS (unmodified, same default params as the Task 2 baseline) on the
# preprocessed + feature-engineered dataset
elcs_preprocessed = eLCS(learning_iterations=10000, random_state=42)
elcs_preprocessed.fit(X_train, y_train)
print("elcs_preprocessed trained")

elcs_preprocessed trained


In [20]:
elcs_improved.fit(X_train, y_train)
print("elcs_improved trained")

elcs_improved trained


In [21]:
# Conventional model (individual contribution): Decision Tree
dt_model = DecisionTreeClassifier(random_state=42, max_depth=8)
dt_model.fit(X_train, y_train)
print("dt_model trained")

dt_model trained


### 5.3 Evaluation metrics

In [22]:
models = {
    "eLCS (preprocessed, original)": elcs_preprocessed,
    "eLCS (improved)": elcs_improved,
    "Decision Tree": dt_model,
}

results = []
predictions = {}
for name, model in models.items():
    y_pred = model.predict(X_test)
    predictions[name] = y_pred
    results.append({
        "model": name,
        "accuracy": accuracy_score(y_test, y_pred),
        "balanced_accuracy": balanced_accuracy_score(y_test, y_pred),
        "f1_macro": f1_score(y_test, y_pred, average="macro"),
    })

results_df = pd.DataFrame(results)
results_df

,model,accuracy,balanced_accuracy,f1_macro
0,"eLCS (preprocessed, original)",0.582367,0.582525,0.496480
1,eLCS (improved),0.582867,0.583000,0.509929
2,Decision Tree,0.624967,0.625015,0.624756


In [23]:
for name, y_pred in predictions.items():
    print(f"--- {name} ---")
    print(classification_report(y_test, y_pred, target_names=["Low", "Medium", "High"]))

--- eLCS (preprocessed, original) ---
              precision    recall  f1-score   support

         Low       0.59      0.84      0.69     10005
      Medium       0.49      0.06      0.11     10006
        High       0.58      0.85      0.69      9989

    accuracy                           0.58     30000
   macro avg       0.55      0.58      0.50     30000
weighted avg       0.55      0.58      0.50     30000

--- eLCS (improved) ---
              precision    recall  f1-score   support

         Low       0.57      0.86      0.69     10005
      Medium       0.48      0.09      0.15     10006
        High       0.61      0.80      0.69      9989

    accuracy                           0.58     30000
   macro avg       0.55      0.58      0.51     30000
weighted avg       0.55      0.58      0.51     30000

--- Decision Tree ---
              precision    recall  f1-score   support

         Low       0.70      0.69      0.70     10005
      Medium       0.48      0.48      0.48  

### 5.4 Statistical significance testing

McNemar's test is used because predictions come from a single shared test set (paired, not independent samples) — the standard test for comparing two classifiers' error patterns on identical instances.

In [24]:
def mcnemar_test(y_true, pred_a, pred_b):
    correct_a = (pred_a == y_true)
    correct_b = (pred_b == y_true)
    table = np.zeros((2, 2))
    table[0, 0] = np.sum(correct_a & correct_b)
    table[0, 1] = np.sum(correct_a & ~correct_b)
    table[1, 0] = np.sum(~correct_a & correct_b)
    table[1, 1] = np.sum(~correct_a & ~correct_b)
    return mcnemar(table, exact=False, correction=True)

In [25]:
pairs = [
    ("eLCS (improved)", "eLCS (preprocessed, original)"),
    ("eLCS (improved)", "Decision Tree"),
]

for a, b in pairs:
    result = mcnemar_test(y_test, predictions[a], predictions[b])
    print(f"{a} vs {b}: statistic={result.statistic:.3f}, p-value={result.pvalue:.4g}")

eLCS (improved) vs eLCS (preprocessed, original): statistic=0.106, p-value=0.7447
eLCS (improved) vs Decision Tree: statistic=227.293, p-value=2.321e-51
